### PR TITLE
`ModelExperimental`: Add basic PBR lighting

### PR DESCRIPTION
**Pull request by ptrgags**
This branch targets on `model-experimental-unlit`. Do not merge until that one is merged.

This PR adds basic PBR lighting (metallic roughness and specular glossiness) to `ModelExperimental`.

Expected results when used with the Pawns model in `/Apps/SampleData/models/Pawns/Pawns.glb`:

image

Note that since this branch does not handle IBL, the gold pawn and the rightmost pawn (which should be silver) do not look correct yet. They will in a future branch.

Image based-lighting is more involved, so I'm splitting that into another branch to keep this one easy to review.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @ptrgags!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

File: Source/Shaders/Builtin/Functions/imageBasedLighting.glsl [hunk 1]
```diff
@@ -0,0 +1,125 @@
+void proceduralIBL(
+  vec3 v,
+  vec3 n,
+  vec3 positionWC,
+  vec3 lightColorHdr,
+  czm_pbrParameters pbrParameters
+) {
```
Comment (sanjeetsuhag) [inline Source/Shaders/Builtin/Functions/imageBasedLighting.glsl:7]: Style - does the curly bracket need to be on the next line?

Comment (ptrgags) [reply Source/Shaders/Builtin/Functions/imageBasedLighting.glsl:7]: Hm, yeah I think it should.

Also... this IBL file shouldn't exist on this branch, I'll remove it

**Comment by ptrgags** [comment]
@sanjeetsuhag updated! though one thing I've noticed is

[this sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=bVPbbtswDP0VwU8JkEpO6/XipsWw9LIBLRY0QffiF8WmY2GyZEhymmzov4+WHTdL4xeLpA55eEhNwYq6pPebCowoQTkuH4C72oCloPhSwrPOQO7HyQ1xpobrRCVqzQ1ZC3gDg14Fb2Ta5nv1vkESpN6eagQKBSYJhj2u0lY4oRUiO9SUG4cnrs5obnR5BysDYAeJIuRkfHpGw4soOh9fjRpHFNHwS3h2EZ57M0xUk7nJW1QHZL4Dz4RazYRLixct5SAckWfuCjr7QRiJRiT8IJWLDWQPhpewMFzZXJvyg1/vslTqFJXy9/RDj3kE7JE7bTznJFDauCIJRq31BtYlQUu0rVY22iIVIzZHqxQHzPdr+RLNt9NxtHOgAP25y3kvpaisFhn99Ti/jPrwkXbb0AFHZNeOmdoUe6QVLgMWXeOW8CzzVLpKn9bFj/JRunzwt01dGxmjGmxeQWrZHXeceZBls28v7BVwBzZTLbVZoF6HNl1hpk5RQhjrkhXOVTFjfiqFti6Ori6u2EouTyRXDoxi3R/xyx6+J3+8b3RhoVIDbRNy+6R5toCNfxkxybm0sONQCiXKupyhlnIu/kBMxqeXPsf7sJ92q16KShtOc7ld6FaPDJsSijcDjP8fZVabzo3L/T68DkbBxLqthNvd9L6KssINaxQdUMoclJXk+ILYsk5/g6OptU315uqE7UMnmVgTkd0ceZ8kldxajOS19N0kwe2E4f1PUKn9cv5cg5F821wrxrdPrZNSOmFoHkc6reWSm4PM/wA) renders the VertexColorTest model correctly

And yet in [your sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=xVdRb9s2EP4rhJ8UwJacxm3X1A3apEkWoEaD2m0epj4w0tkmSpECScnKhvz3HklJVj1tg9u5BWyTd+Td9x1JiZ9LqkjJYAOKvCICNuQCNCuy8JPzBfEgcfaFFIYyASoeHL2MRSyiiMxkCpzcSm2YFEysYlFislxqZh2Yrk51QZXBHhUn4VLJ7C2sFIAOYkHI6PjJSTh+Ppk8O34xtI7JJBw/HZ88Hz9z5tNxLCyeTbzOdyj+DjRF2FtmkvUHyXkwHhL7aQKWrIL0StEMFooKvZQq25JqXTrkMqHcz5NXbcw1YLXUSOWIxgMhlVnHg6G3NqBNPPDkusuhPXTm+oj2h53+l/0hxEBlTjH2XMrsXFZ1LkKk0MAhwbFlIRK3dsFRE0SIH3TZg8ZnKUTzHBIdvaWGRh47uj3/ENXZmzZccbO0TH2Y5WvbR9s8DnvZVZcZ05qVcDCGLUK3/z1MPwrOzOFouvRtZ5QriU+DhvR7qN4IbahIID0A3Wuk807SFFTURYpWfHEV3ctqxBrXDzF3zwynlunPK6ID2lPPyGyH/6/a7phZz5iY0eqXVLmF/496RxnDL61+qO4bYUBxoOXPPJgd0L4a2XZ479pmLMGnVOrssNW0MJ5/1ph7810ovBhXHOyuy8LciJThO+aw5PsxfSX9Y/uXtZFzlkJ6y6k41D3yDca31t58PwFqlOpCcqkW7mo/COMdlF17b9Z3QM3aarJDnpYaxB+PjTf+lWksPjtV1HLo4haK14y86gx1gkIrzBXLUDaWeNQUZLKEN6jnfNpWT6Gc+qcYmqZ1RbW8c2CXVQ44B1C6cic9bVVBuxxI5dT+DBsHEwmCu+n8wVa/wHUuFGhcT8o1tBMdnRk1ilWnPYJyvSNMu7Kys/CNUh5uXahxO1ad+ZJzlmvJ0vDuev7bpDOhR+G2W1LviGv9QnYWPcEQRcMlf1jIc1kIS3ee48ZC0BHYOyMtX3I8Pho2xyotlLuNTsnY7b7FevSieE5FmlBtONgNWkjJ76magSgCr4+PXg6Gg6k2DxzOGt6vWZaj0rb7EoRhZCDL8bbDt9N9kXwBEyZaN6dtGnVDpykrCUtf9fxjIQmnWuPIsuB8zv7EF9LZNML5fwvl0u3c+xLvH/pgp62Pz955ZxiG0wjN/kjjq9vJ/BU) it renders but I don't see anything on the screen...

File: Source/Shaders/ModelExperimental/LightingStageFS.glsl [hunk 2]
```diff
@@ -7,16 +7,75 @@ vec3 LINEARtoSRGB(vec3 linearIn)
...
+
+    vec3 color = czm_pbrLighting(
+        v_positionEC,
+        inputMaterial.normal,
+        czm_lightDirectionEC,
+        lightColorHdr,
+        pbrParameters
+    );
+
+    // TODO: what are the parameters for IBL?
```
Comment (sanjeetsuhag) [inline Source/Shaders/ModelExperimental/LightingStageFS.glsl:55]: Clean up the IBL comments here.

File: Source/Shaders/ModelExperimental/LightingStageFS.glsl [hunk 3]
```diff
@@ -7,16 +7,75 @@ vec3 LINEARtoSRGB(vec3 linearIn)
     #endif 
 }
 
+#ifdef HAS_OUTLINES
```
Comment (sanjeetsuhag) [inline Source/Shaders/ModelExperimental/LightingStageFS.glsl:10]: We're not setting this flag anywhere, should this be part of a later PR?

**Comment by sanjeetsuhag** [comment]
> @sanjeetsuhag updated! though one thing I've noticed is
> 
> [this sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=bVPbbtswDP0VwU8JkEpO6/XipsWw9LIBLRY0QffiF8WmY2GyZEhymmzov4+WHTdL4xeLpA55eEhNwYq6pPebCowoQTkuH4C72oCloPhSwrPOQO7HyQ1xpobrRCVqzQ1ZC3gDg14Fb2Ta5nv1vkESpN6eagQKBSYJhj2u0lY4oRUiO9SUG4cnrs5obnR5BysDYAeJIuRkfHpGw4soOh9fjRpHFNHwS3h2EZ57M0xUk7nJW1QHZL4Dz4RazYRLixct5SAckWfuCjr7QRiJRiT8IJWLDWQPhpewMFzZXJvyg1/vslTqFJXy9/RDj3kE7JE7bTznJFDauCIJRq31BtYlQUu0rVY22iIVIzZHqxQHzPdr+RLNt9NxtHOgAP25y3kvpaisFhn99Ti/jPrwkXbb0AFHZNeOmdoUe6QVLgMWXeOW8CzzVLpKn9bFj/JRunzwt01dGxmjGmxeQWrZHXeceZBls28v7BVwBzZTLbVZoF6HNl1hpk5RQhjrkhXOVTFjfiqFti6Ori6u2EouTyRXDoxi3R/xyx6+J3+8b3RhoVIDbRNy+6R5toCNfxkxybm0sONQCiXKupyhlnIu/kBMxqeXPsf7sJ92q16KShtOc7ld6FaPDJsSijcDjP8fZVabzo3L/T68DkbBxLqthNvd9L6KssINaxQdUMoclJXk+ILYsk5/g6OptU315uqE7UMnmVgTkd0ceZ8kldxajOS19N0kwe2E4f1PUKn9cv5cg5F821wrxrdPrZNSOmFoHkc6reWSm4PM/wA) renders the VertexColorTest model correctly
> 
> And yet in [your sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=xVdRb9s2EP4rhJ8UwJacxm3X1A3apEkWoEaD2m0epj4w0tkmSpECScnKhvz3HklJVj1tg9u5BWyTd+Td9x1JiZ9LqkjJYAOKvCICNuQCNCuy8JPzBfEgcfaFFIYyASoeHL2MRSyiiMxkCpzcSm2YFEysYlFislxqZh2Yrk51QZXBHhUn4VLJ7C2sFIAOYkHI6PjJSTh+Ppk8O34xtI7JJBw/HZ88Hz9z5tNxLCyeTbzOdyj+DjRF2FtmkvUHyXkwHhL7aQKWrIL0StEMFooKvZQq25JqXTrkMqHcz5NXbcw1YLXUSOWIxgMhlVnHg6G3NqBNPPDkusuhPXTm+oj2h53+l/0hxEBlTjH2XMrsXFZ1LkKk0MAhwbFlIRK3dsFRE0SIH3TZg8ZnKUTzHBIdvaWGRh47uj3/ENXZmzZccbO0TH2Y5WvbR9s8DnvZVZcZ05qVcDCGLUK3/z1MPwrOzOFouvRtZ5QriU+DhvR7qN4IbahIID0A3Wuk807SFFTURYpWfHEV3ctqxBrXDzF3zwynlunPK6ID2lPPyGyH/6/a7phZz5iY0eqXVLmF/496RxnDL61+qO4bYUBxoOXPPJgd0L4a2XZ479pmLMGnVOrssNW0MJ5/1ph7810ovBhXHOyuy8LciJThO+aw5PsxfSX9Y/uXtZFzlkJ6y6k41D3yDca31t58PwFqlOpCcqkW7mo/COMdlF17b9Z3QM3aarJDnpYaxB+PjTf+lWksPjtV1HLo4haK14y86gx1gkIrzBXLUDaWeNQUZLKEN6jnfNpWT6Gc+qcYmqZ1RbW8c2CXVQ44B1C6cic9bVVBuxxI5dT+DBsHEwmCu+n8wVa/wHUuFGhcT8o1tBMdnRk1ilWnPYJyvSNMu7Kys/CNUh5uXahxO1ad+ZJzlmvJ0vDuev7bpDOhR+G2W1LviGv9QnYWPcEQRcMlf1jIc1kIS3ee48ZC0BHYOyMtX3I8Pho2xyotlLuNTsnY7b7FevSieE5FmlBtONgNWkjJ76magSgCr4+PXg6Gg6k2DxzOGt6vWZaj0rb7EoRhZCDL8bbDt9N9kXwBEyZaN6dtGnVDpykrCUtf9fxjIQmnWuPIsuB8zv7EF9LZNML5fwvl0u3c+xLvH/pgp62Pz955ZxiG0wjN/kjjq9vJ/BU) it renders but I don't see anything on the screen...
> 
> It's not the indexing this time, the unlit cube renders just fine.

It works, but you have the move the camera around to see the front face.

image

File: Source/Shaders/ModelExperimental/LightingStageFS.glsl (hunk 3 above)
Comment (ptrgags) [reply Source/Shaders/ModelExperimental/LightingStageFS.glsl:10]: ah yeah, this should be done later when we get to outlines for Cesium OSM buildings. I'll remove it for now.

**Comment by ptrgags** [comment]
@sanjeetsuhag interesting, this time it failed with timeouts again. Since I commented out the actual MaterialPipelineStage code, I think this means it has something to do with how I'm using loadGltf or resource management thereof. We can talk about this more tomorrow morning.

**Comment by lilleyse** [comment]
@ptrgags is https://github.com/CesiumGS/cesium/issues/9710 related? Those timeouts are showing up in `main` as well. Probably my fault but you or @sanjeetsuhag may need to investigate.

**Comment by ptrgags** [comment]
@lilleyse hm that is the same type of error that we're seeing. Our unit tests were modeled after the GltfLoader one so whatever the fix there is, it's probably applicable here too... We should discuss this tomorrow

**Comment by sanjeetsuhag** [comment]
@lilleyse I see that `waitForLoaderProcess` only waits for the `promise` and not the `texturesLoadedPromise`. Do you think that's playing a part here?

https://github.com/CesiumGS/cesium/blob/ffcc025a9ed8c9f54e7d7ce6cb09bf89bcdac273/Specs/waitForLoaderProcess.js#L3-L14

**Comment by lilleyse** [comment]
@sanjeetsuhag let's move the discussion to https://github.com/CesiumGS/cesium/pull/9712 so this PR isn't blocked.

**Comment by ptrgags** [comment]
@sanjeetsuhag reverted the commit that commented out test bodies, this is ready for final review

**Comment by ptrgags** [comment]
@sanjeetsuhag force-pushed an update that fixes the syncing with `model-experimental-staging`

File: Source/Shaders/ModelExperimental/LightingStageFS.glsl [hunk 4]
```diff
@@ -7,16 +7,59 @@ vec3 LINEARtoSRGB(vec3 linearIn)
...
+}
+
+vec3 computePbrLighting(czm_modelMaterial inputMaterial)
+{
+    czm_pbrParameters pbrParameters;
+    pbrParameters.diffuseColor = inputMaterial.diffuse;
+    pbrParameters.f0 = inputMaterial.specular;
+    pbrParameters.roughness = inputMaterial.roughness;
+
+    #ifndef USE_CUSTOM_LIGHT_COLOR
```
Comment (sanjeetsuhag) [inline Source/Shaders/ModelExperimental/LightingStageFS.glsl:27]: We're currently not passing in a `lightColor` to `ModelExperimental` nor are we setting this define anywhere.

Comment (sanjeetsuhag) [reply Source/Shaders/ModelExperimental/LightingStageFS.glsl:27]: Spoke to @lilleyse - we can add this feature at a later time.

Comment (ptrgags) [reply Source/Shaders/ModelExperimental/LightingStageFS.glsl]: Removed the unused light color for now.

**Comment by sanjeetsuhag** [comment]
The code looks good and the models look as expected. The CI failures will be handled in #9710, so I'm merging. Thanks @ptrgags!

**Pull request merged**